File: apps/admin/src/models/Regularization.ts
import mongoose, { Document, Schema } from 'mongoose';
import type { RegularizationStatus, RegularizationType } from '@app-types/enums';

export interface IRegularization extends Document {
  employeeId: mongoose.Types.ObjectId;
  dateString: string;
  type: RegularizationType;
  requestedCheckIn?: Date;
  requestedCheckOut?: Date;
  reason: string;
  status: RegularizationStatus;
  reviewedBy?: mongoose.Types.ObjectId;
  reviewedAt?: Date;
  reviewRemark?: string;
  attendanceDayId?: mongoose.Types.ObjectId;
  withdrawnAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

const RegularizationSchema = new Schema<IRegularization>(
  {
    employeeId:       { type: Schema.Types.ObjectId, ref: 'User', required: true },
    dateString:       { type: String, required: true, match: /^\d{4}-\d{2}-\d{2}$/ },
    type:             { type: String, enum: ['forgotCheckIn','forgotCheckOut','workAwayFromOffice','officialTravel','clientVisit'], required: true },
    requestedCheckIn:  { type: Date },
    requestedCheckOut: { type: Date },
    reason:           { type: String, required: true, maxlength: 500 },
    status:           { type: String, enum: ['pending','approved','rejected','withdrawn'], default: 'pending' },
    reviewedBy:       { type: Schema.Types.ObjectId, ref: 'User' },
    reviewedAt:       Date,
    reviewRemark:     String,
    attendanceDayId:  { type: Schema.Types.ObjectId, ref: 'AttendanceDay' },
    withdrawnAt:      Date,
  },
  { timestamps: true },
);

// One active (pending/approved) request per employee per date
RegularizationSchema.index({ employeeId: 1, dateString: 1, status: 1 });
RegularizationSchema.index({ employeeId: 1, status: 1 });
RegularizationSchema.index({ status: 1, createdAt: -1 });

export const Regularization =
  mongoose.models.Regularization ??
  mongoose.model<IRegularization>('Regularization', RegularizationSchema);

File: apps/mobile/lib/core/models/regularization.dart
class RegularizationRequest {
  final String id;
  final String employeeId;
  final String date;
  final String type;
  final String reason;
  final String status;
  final String? requestedCheckIn;
  final String? requestedCheckOut;
  final String? reviewedBy;
  final String? reviewedAt;
  final String? reviewRemark;
  final String createdAt;

  const RegularizationRequest({
    required this.id,
    required this.employeeId,
    required this.date,
    required this.type,
    required this.reason,
    required this.status,
    this.requestedCheckIn,
    this.requestedCheckOut,
    this.reviewedBy,
    this.reviewedAt,
    this.reviewRemark,
    required this.createdAt,
  });

  bool get isCancellable => status == 'pending';
  bool get isWfh => type == 'wfh';
  bool get isMissedPunch => type == 'missed_punch';

  factory RegularizationRequest.fromJson(Map<String, dynamic> json) {
    return RegularizationRequest(
      id: json['_id'] as String? ?? '',
      employeeId: _extractId(json['employeeId']),
      date: json['date'] as String? ?? '',
      type: json['type'] as String? ?? '',
      reason: json['reason'] as String? ?? '',
      status: json['status'] as String? ?? 'pending',
      requestedCheckIn: json['requestedCheckIn'] as String?,
      requestedCheckOut: json['requestedCheckOut'] as String?,
      reviewedBy: _extractIdNullable(json['reviewedBy']),
      reviewedAt: json['reviewedAt'] as String?,
      reviewRemark: json['reviewRemark'] as String?,
      createdAt: json['createdAt'] as String? ?? '',
    );
  }

  static String _extractId(dynamic value) {
    if (value is String) return value;
    if (value is Map) return value['_id'] as String? ?? '';
    return '';
  }

  static String? _extractIdNullable(dynamic value) {
    if (value == null) return null;
    if (value is String) return value;
    if (value is Map) return value['_id'] as String?;
    return null;
  }
}
